### ISSUE TITLE
Cypress not playing great with Newrelic

### ISSUE THREAD

**Issue by AvnerCohen** [cypress-io/cypress#9058]
### Current behavior
Javascript error is thrown when using Cypress together with Newrelic agent running on the browser.
Error is:
```
Unable to set property 'status' of undefined error
```
### Desired behavior
No error

### Test code to reproduce
This is not so much about the code and more about the integration with newrelic
In our case, we use Pyramid server (python) with the latest newrelic agent.

### Versions
5.5

### Further information
This seems to be a known issue on the newrelic end:
https://discuss.newrelic.com/t/exception-in-nrwrapper/72759/18
I recently created another issue - https://discuss.newrelic.com/t/playing-better-with-cypress/120046

And also, seems like people are trying to resolve this on cypress end - spotted in the wild:
https://github.com/BYScribble/ion-e2e-samples/blob/739489ce17b2bbbed05ebb949a8fa677762ee3b1/cypress/support/disableNewRelic.js
https://github.com/psincraian/pepy-front/blob/1ed98a5d3060947501cef31713e66380fa993a51/cypress/support/index.js#L24

Trying all solutions, I couldn't get this to work, and I am wondering if maybe there is a way to fix the root cause here, the conflict itself, rather than fixing after the fact or ignoring the error?

**Comment by jennifer-shehane** [comment]
I would say that all of these are pretty solid workarounds for NewRelic throwing an uncaught exception. The exact code around them may be older though.  

1. Using the `uncaught:exception` handler and specifically regexing out the NewRelic error will still cause Cypress to fail tests on other uncaught exceptions (this error message construction changed in 5.0, so this may need updating): https://github.com/psincraian/pepy-front/blob/1ed98a5d3060947501cef31713e66380fa993a51/cypress/support/index.js#L24
2. Replacing the script on load of the page (also, did not check this exact code, but in theory this is possible): https://github.com/BYScribble/ion-e2e-samples/blob/739489ce17b2bbbed05ebb949a8fa677762ee3b1/cypress/support/disableNewRelic.js
3. Ideally I would like it if Cypress didn't fail the test on uncaught:exceptions thrown as outlined here: https://github.com/cypress-io/cypress/issues/4933 so I would suggest following along there as well as the larger delivery of displaying warning message in the Command Log: https://github.com/cypress-io/cypress/issues/4917

**Comment by AvnerCohen** [comment]
@jennifer-shehane I really appreciate your response.

As you may have noticed, I did review quite a few potential fixes to this problem and came up empty and concluding that there is no satisfactory solution at this point, hence, I tend to disagree and I will explain:

1. While this indeed prevents Cypress from failing the test by silencing the message, the error itself is still thrown and so, the results is many Ks of error messages sent to newrelic, outside of noise (and costs) it also impact performances of the testing, so this is not a real silencing of the issue.

2. This is a nice attempt, based on https://github.com/bahmutov/stop-script, but this is not working in many cases, why ? because the newrelic script is not injected on the client side, the full script is sent from server side, I have provided the link to the (for expample) python agent that is doing that.

3. I guess I am missing something here. Personally, I am not sure what the better behavior for uncaught:exceptions should be, the point I am raising in this ticket is that I would like to see the *root cause*, the reason why:
`Unable to set property 'status' of undefined error`
Is thrown to begin with, being fixed.

Maybe it's a naming conflict? some global pollution ? I am happy to try and deep dive with some pointers.

At any case, if you feel despite this, the cypress position at this point is that there is not enough need or demand to spend any time on this, being a very specific newrelic<>cypress thing, please let me know and I can mark this in my head as `won't fix`.

**Comment by jennifer-shehane** [comment]
@AvnerCohen K, sorry, I originally thought this was an issue with NewRelic alone, but I know have reviewed this thread and see how this is an issue uniquely when NewRelic is run within Cypress. https://discuss.newrelic.com/t/exception-in-nrwrapper/72759/9 Someone from NewRelic seemed to explain:

>Cypress.io 12 patches XMLHttpRequest, similarly to the NR agent. However, it modifies the prototype chain, so that the open and send methods cannot be instrumented by us. This has a side-effect of the agent causing an error when tests are run in Cypress. This should not affect production apps, but it will affect customers who are running tests with the NR agent present in their code.
>https://discuss.newrelic.com/t/exception-in-nrwrapper/72759/10

I think it's notable that this is not only a conflict with NewRelic and Cypress, but people in the thread mention this error occuring due to a conflict with NewRelic and Bugsnap and Rollbar. 

@AvnerCohen What is the use case for using NewRelic's Browser Monitoring within Cypress? Is there a use case for collecting data while running tests? Just wondering if you'd rather it not report at all when within Cypress.

**Comment by AvnerCohen** [comment]
@jennifer-shehane thanks so much for deep diving here !
The use case is a smoke testing running over production environment (where new relic is installed) after a version deploy.
My failed attempt was to eliminate new-relic all together, so I am totally fine with it not collecting anything during these runs.

As I wrote, for that, I also tried to use the blockHost cypress config, but this too failed, because this is too late in the process the issue is not about API requests to newrelic, it's about the first script present on the page as part of the initial response from server.

**Comment by jennifer-shehane** [comment]
Can you detect you're running your app within Cypress and change how you're using NewRelic within testing?

So adding something like below to your application code?

```js
if (window.Cypress) {
  // don't send NewRelic data
} else {
  // send NewRelic data
}
```

**Comment by AvnerCohen** [comment]
@jennifer-shehane I wish :)

The newrelic server side JS instrumentation works like that, for example:

```html
<!doctype html>
<html>
<head>
  <title>Awesome Login Page</title>
  <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/>
  <link href='//fonts.googleapis.com/css?family=Roboto:400,700,300,500' rel='stylesheet' type='text/css'>
....

```

Will change into this served template:

```html

<!doctype html>
  <html>
  <head>
  <title>Awesome Login Page</title>
  <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/>
   <script type="text/javascript">(window.NREUM\|\|(NREUM={})).loader_config= .... 
   <link href='//fonts.googleapis.com/css?family=Roboto:400,700,300,500' rel='stylesheet' type='text/css'>
    ....

```

This happens in the agent code here - https://github.com/newrelic/newrelic-python-agent/blob/77e000cf4fd04a5d0921e334ff34d3b3be2be008/newrelic/api/html_insertion.py#L32
With zero options to custom wrap this string (it's an empty string or full string on the server level):
https://github.com/newrelic/newrelic-python-agent/blob/77e000cf4fd04a5d0921e334ff34d3b3be2be008/newrelic/api/web_transaction.py#L351


You can see more on how it's actually done in here - https://docs.newrelic.com/docs/agents/python-agent/installation/standard-python-agent-install
Basically, the web server executable is being started with:
```sh
newrelic-admin $WEB_SERVER NAME
```
And this allows the new relic instrumentation to happen automagically, so very little room for manual interference there.

**Comment by rbayliss** [comment]
This issue is very much affecting us, too.  It used to be the spurious `Unable to set property 'status' of undefined error` issue, but since we've switched to using the new experimental network stubbing (`route2`) in Cypress, we've noticed that our application breaks when it is rerun in the same Cypress browser session. The issue there is similar - an uncaught error, this time only appearing during ajax HTTP requests made on retry:
```
nr-spa-1167.min.js:1 Uncaught (in promise) TypeError: Cannot read property 'method' of undefined
```
Tracing it back, it also appears to be related to things New Relic expects to be present on the request or response object that aren't there.  The silver lining is that we can pretty easily use `cy.route2` to block New Relic from loading on the page:
```
// Block new-relic.js outright due to issues with Cypress networking code.
cy.route2("**/new-relic.js", (req) => {
  req.reply("console.log('Fake New Relic script loaded');");
});
```

**Comment by acornellier** [comment]
I'm have the same issue. Getting around it by intercepting the request to new relic doesn't seem to work. And using `Cypress.on('uncaught:exception'` is not an elegant solution as the errors still go to the console. Not sure how to progress.

**Comment by mvaccarouy** [comment]
I got it to work with this:

`before(() => {
    // Block newrelic js outright due to issues with Cypress networking code.
    cy.log("Blocking NewRelic scripts");

    //Will block 
    //  https://js-agent.newrelic.com/nr-spa-1208.js
    cy.intercept(/\.*newrelic.*$/, (req) => {
        console.log("NEW RELIC INTERCEPTED");
        req.reply("console.log('Fake New Relic script loaded');");
    });

});`

**Comment by lucksp** [comment]
The solution above doesn't work for us:


I have it implemented as a repeatable custom command:
```
Cypress.Commands.add('interceptNewRelic', () => {
  cy.log("Blocking NewRelic scripts");

    // Will block 
    // https://js-agent.newrelic.com/**.js
    cy.intercept(/\.*newrelic.*$/, (req) => {
      console.log("NEW RELIC INTERCEPTED");
      req.reply("console.log('Fake New Relic script loaded');");
    });
})
```

I also tried:

```
Cypress.Commands.add('interceptNewRelic', () => {
     cy.on('uncaught:exception', (err, runnable) => {
      cy.log({err});
      // returning false here prevents Cypress from
      // failing the test
      return false
    })
})
```
<img width="438" alt="image" src="https://user-images.githubusercontent.com/15162169/210889075-4e8adc32-39e6-4425-839a-7527bdffe9b9.png">
